### PR TITLE
[WIP] Add an :ets table to cache the list of files that can be formatted

### LINKED ISSUE #385: Eliixir LS slow since upgrading to elixir 1.11
Hiya,

Ever since ive upgraded to Elixir 1.11, elixir ls has become quite slow, formatting a file, even if the changes are just some spaces can take over 30 sec.

### Environment

* Elixir & Erlang versions (elixir --version): 
`Erlang/OTP 23 [erts-11.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1] [hipe]`
`Elixir 1.11.0 (compiled with Erlang/OTP 23)`
* Operating system: `MacOS 10.15.7`
* Editor or IDE name (e.g. Emacs/VSCode): `VsCode 1.50.0`
* Editor Plugin/LSP Client name: `v.0.6.1`

### ElixirLS logs
```
[Info  - 8:06:51 AM] [ElixirLS Dialyzer] Checking for stale beam files
[Info  - 8:06:51 AM] [ElixirLS WorkspaceSymbols] Updating index...
[Info  - 8:07:05 AM] [ElixirLS WorkspaceSymbols] 1 modules need reindexing
[Info  - 8:07:05 AM] [ElixirLS WorkspaceSymbols] 0 callbacks added to index
[Info  - 8:07:05 AM] [ElixirLS WorkspaceSymbols] 1 modules added to index
[Info  - 8:07:05 AM] [ElixirLS WorkspaceSymbols] 0 types added to index
[Info  - 8:07:05 AM] [ElixirLS WorkspaceSymbols] 10 functions added to index
[Info  - 8:07:30 AM] [ElixirLS Dialyzer] Found 2760 changed files in 20172 milliseconds
[Info  - 8:08:24 AM] [ElixirLS Dialyzer] Analysis finished in 314742 milliseconds
```

### PR DESCRIPTION
**Pull request by benwilson512**
# Cached Input List

This PR improves `FormatOnSave` time by several orders of magnitude, at least when tested on my PC. 

Before: 900ms
After: ~2-3ms

An O(n), file system bottlenecked operation is replaced by an O(1) in memory operation.

## Related
Fixes #385.

## Problem

As noted in #385, #381 and reflected in my experience with large projects at CargoSense, FormatOnSave could cause significant to massive delays when trying to save files. After some hunting, the vast majority of the time is spent here:

https://github.com/elixir-lsp/elixir-ls/blob/master/apps/language_server/lib/language_server/providers/formatting.ex#L50

The job of this function is to figure out if a specific file that is going to be saved should be formatted. The key factor here is the `:inputs` criterion defined in `.formatter.exs` which takes the shape of a list of path globs eg:

```
inputs: [
    "*.exs",
    "config/**/*.exs",
    "apps/*/{config,lib,test}/**/*.{ex,exs}",
    "apps/*/mix.exs"
  ]
```

The current approach simply calls `Path.wildcard` on each entry, expanding it into all files based on that glob. Then, the candidate path is tested for membership in that list.

## Proposed Approach

This PR does NOT change the overall game plan. There is still a list of files which is produced by calling `Path.wildcard` on the input list, and a candidate path is still tested for membership. However, the main insight here is that this list very rarely changes. This PR populates that list when the LanguageServer boots and caches it in an `:ets` table set. Then, testing membership becomes an extremely fast process, just an `:ets.lookup` call. These `:ets` calls _are_ serialized through the GenServer to avoid various race conditions, but they are such a small percentage of the overall format time that I'm not worried about it being a bottleneck.

TODO:
- [x] Invalidate the cache when `.formatter.exs` changes.
- [x] Block formatting until the cache has been initialized so efforts to format don't crash.
- [x] Block formatting while the cache is being rebuilt.
- [x] Replace IO.puts with whatever the proper logging thing is.
- [x] Handle deletion / creation of formattable files
- [x] Handle deletion / creation of `.formatter.exs`
- [ ] Fix tests.
- [x] Fix "can't find dep" issues.
- [ ] Validate that `cd`ing into an umbrella sub-dir works.

### CODE REVIEW

File: apps/language_server/lib/language_server/providers/formatting.ex [hunk 1]
```diff
@@ -103,4 +108,63 @@ defmodule ElixirLS.LanguageServer.Providers.Formatting do
...
+  end
+
+  def handle_call({:build_cache, dir}, _, state) do
+    opts_result = SourceFile.formatter_opts(dir)
+    :ets.delete_all_objects(__MODULE__)
+    :ets.insert(__MODULE__, {:formatter_opts, opts_result})
+
+    case opts_result do
+      {:ok, opts} ->
+        IO.puts("[ElixirLS FormattingCache] Formatter building cache")
```
Comment (benwilson512) [inline apps/language_server/lib/language_server/providers/formatting.ex:134]: I tried using the same function as here https://github.com/elixir-lsp/elixir-ls/blob/master/apps/language_server/lib/language_server/providers/workspace_symbols.ex#L130 but got some sort of process error. Will improve.

**Comment by benwilson512** [comment]
Any ideas about the test failures? They're getting "Process not found" errors, but the process is in the app supervision tree. Is there some spot in the test setup I should be adding it to?

**Comment by lukaszsamson** [comment]
Nice @benwilson512 . I've got some points
1. How about caching formatter opts along with formattable files?
I guess it will also need another TODOs:
2. Invalidate cache when `.formatter.exs` files are created/deleted
3. Invalidate cache when formattable files are created/deleted/renamed

**Comment by benwilson512** [comment]
> How about caching formatter opts along with formattable files?
Already done!

As for the others, great point! Need to find the right hooks in the LanguageServer to identify these cases.

File: apps/language_server/lib/language_server/server.ex [hunk 2]
```diff
@@ -320,6 +322,14 @@ defmodule ElixirLS.LanguageServer.Server do
     |> Enum.uniq()
     |> WorkspaceSymbols.notify_uris_modified()
 
+    formatter_changed? = Enum.any?(changes, fn %{"uri" => uri} ->
+      Path.basename(uri) == ".formatter.exs"
```
Comment (benwilson512) [inline apps/language_server/lib/language_server/server.ex:326]: Is it more complicated than this? I realize this could lead to some false positives but that isn't really a big issue IMHO. Are there files beyond this file that can be allowed to affect formatting?

**Review by benwilson512 — COMMENTED** [review]
@lukaszsamson OK I believe those scenarios are all handled now.

File: apps/language_server/lib/language_server/server.ex (hunk 2 above)
Comment (lukaszsamson) [reply apps/language_server/lib/language_server/server.ex:326]: How about `.formatter.exs` in subdirectories? AFAIK it is supported in subapps in umbrella and there's an unfinished PR that tries to make it work in more cases https://github.com/elixir-lsp/elixir-ls/pull/362
Though editing `.formatter.exs` is likely rare I can see that happen more often when switching between branches

Comment (benwilson512) [reply apps/language_server/lib/language_server/server.ex:326]: @lukaszsamson that will work just fine here. What this code is doing is saying "If any files are changed anywhere in the open project named ".formatter.exs" then rebuild the cache.

I do suppose the following would be an issue: If you `cd` into an elixir project's `lib` folder and `code .` there, it won't be watching the `.formatter.exs` in the upper directory. Let me see if that is honored today though, if not then this at least isn't a regression.

Comment (lukaszsamson) [reply apps/language_server/lib/language_server/server.ex:326]: OK, I didn't notice ``Path.basename` call.

Regarding the `cd` to `lib` - we don't support non mix projects.

Comment (LostKobrakai) [reply apps/language_server/lib/language_server/server.ex:326]: `cd` into `lib` might be unsupported, but what about `cd` into `apps/myapp` within an umbrella? A single app within an umbrella is still considered a mix project afaik.

Comment (benwilson512) [reply apps/language_server/lib/language_server/server.ex:326]: Awesome, that definitely simplifies things. There are still a few bugs I'm working out, hope to have this in a ready state by early next week.

**Comment by axelson** [comment]
Regarding the `Fix "can't find dep" issues` TODO item I have some local work that I believe can alleviate that pain point (it works by copying over some logic from `Mix.Tasks.Formatter.formatter_opts_for_file` into ElixirLS, but I hope in the future that could be upstreamed to Elixir). So maybe that should be removed as a goal for this PR and it can be implemented separately, that might be a good idea even if we don't end up going with my approach, to keep the size and feasibility of this PR down.

**Comment by benwilson512** [comment]
@axelson I mean the issue is that at the moment in this PR dep importing is _entirely broken_ so we really can't merge it in. Hopefully I'll get a chance to fix that in this branch soon.

**Comment by axelson** [comment]
Ah, I see. In that case it is important to handle within this PR :+1:

**Comment by axelson** [comment]
@benwilson512 ping :)

Have you had a chance to look at this recently? It would be great to be get it in.

**Comment by benwilson512** [comment]
Hey folks! Life at house Wilson has gotten a bit exciting with the introduction of my first child a week ago! Although sleep is a bit hard to find I have taken an extended period off of work and hope to have time to mess with this over the holidays.

**Comment by axelson** [comment]
Congratulations on your first child! I hope you can find some time to enjoy the holidays between everything :)

File: apps/language_server/lib/language_server/server.ex [hunk 3]
```diff
@@ -208,6 +208,7 @@ defmodule ElixirLS.LanguageServer.Server do
 
     if reason == :normal do
       WorkspaceSymbols.notify_build_complete()
+      Formatting.build_cache(state.project_dir)
```
Comment (benwilson512) [inline apps/language_server/lib/language_server/server.ex:211]: OK so this solves the issue with formatting not being able to pull in deps initially. By re-building the the cache after every build we trivially catch updates to deps or other changes that effect formatting.

**Comment by benwilson512** [comment]
I can confirm that this isn't really working at all in umbrella apps, so working on that.

**Comment by billylanchantin** [comment]
@benwilson512 Some problems with the tests may have solved by https://github.com/elixir-lsp/elixir-ls/pull/488. Try updating to see?

**Comment by dsnipe** [comment]
@benwilson512 You made a great addition to ElixirLS and I hope you're able to finish it. But if you don't have time, is it fine if I pick it up? 
I really want to have a fix of the formatting bug in umbrella applications (which is part of this PR). It will make my (and many others) life much easier :)

**Comment by benwilson512** [comment]
Hey @dsnipe feel free to take a swing at it. What's the best way to do that? Would it work to make a PR into this branch? I'd definitely be responsive to that PR and its review, I just didn't have time to chase down the test failures or handle some of the various edge cases like "what if this vscode elixir instance has been opened not in a mix project?".

**Comment by dsnipe** [comment]
@benwilson512 Great, then, I'll try to allocate some time and see what I can do.

**Comment by wingyplus** [comment]
Hi, all. I've start resolving test error step by step. I start open PR to resolve the compiling error here https://github.com/benwilson512/elixir-ls/pull/1 (it's PR against this PR).

**Comment by martasd** [comment]
Hi @benwilson512. What is the status on this one right now?

**Comment by benwilson512** [comment]
I wasn't able to get tests to pass for umbrellas, and haven't had time to followup. If someone wants to take over from here that'd be very appreciated.

**Comment by wingyplus** [comment]
I has been sending a pr to start tackle the test suite here https://github.com/benwilson512/elixir-ls/pull/1.  I’m not sure someone have take a look?

btw, it seem pr got merge conflict. I think we resolve it first before do anything.

**Comment by wingyplus** [comment]
I’m would like to take over it if you’re ok. :)

**Comment by benwilson512** [comment]
Sure, your best bet is to simply fork my branch into your own repository and then open an independent PR. When you've done that I'll close this one.

**Comment by lukaszsamson** [comment]
@benwilson512 @wingyplus  is it still relevant? How's the speed after recent path glob changes?

**Comment by tomjakubowski** [comment]
> The job of this function is to figure out if a specific file that is going to be saved should be formatted. The key factor here is the :inputs criterion defined in .formatter.exs which takes the shape of a list of path globs eg:

```
inputs: [
    "*.exs",
    "config/**/*.exs",
    "apps/*/{config,lib,test}/**/*.{ex,exs}",
    "apps/*/mix.exs"
  ]
```

> The current approach simply calls Path.wildcard on each entry, expanding it into all files based on that glob. Then, the candidate path is tested for membership in that list.

Forgive my drive-by comment here: would it work to check the candidate path against the set of globs without expanding them? If so then there would be no need to cache the list of files at all.

**Comment by benwilson512** [comment]
> couldn't it work just as well to check the candidate path against the set of globs without expanding them?

There is no function in the Elixir standard library to do this that I know of. If there's a well defined spec or library to compare a candidate path against a glob then we could use that.

I think realistically there's no chance at this point that I'm going to finish this, so I'm gonna close the PR.

**Comment by lukaszsamson** [comment]
> > couldn't it work just as well to check the candidate path against the set of globs without expanding them?
> 
> There is no function in the Elixir standard library to do this that I know of. If there's a well defined spec or library to compare a candidate path against a glob then we could use that.

A basic glob match function should be fairly easy to write. At first look glob seems like a simplified regex. The complex part is dealing correctly with cwd, symlinks, special paths, case insensitive filesystems and other path oddities

**Comment by benwilson512** [comment]
> The complex part is dealing correctly with cwd, symlinks, special paths, case insensitive filesystems and other path oddities

Indeed.

**Comment by lukaszsamson** [comment]
Anyway, I forgot that we are already using a simple function like that provided by PathGlob package
